engine/src/main/resources/assets/shaders/chunk_frag.glsl

```diff
@@ -95,34 +101,29 @@ void main() {
 #if defined (NORMAL_MAPPING) || defined (PARALLAX_MAPPING)
     // TODO: Calculates the tangent frame on the fly - this is absurdly costly... But storing
     // the tangent for each vertex in the chunk VBO might be not the best idea either.
-    // Screen-space coordinates happen to be a coordinate system to which it's possible to relate both view-space and texture coordinates, but are themselves irrelevant.
+    // The specific relationship between screen coordinates and view coordinates is irrelevant here. Screen coordinates just happen to be a basis that it's possible to relate both view and UV space coordinates to.
     mat2x3 screenToView = mat2x3(dFdx(vertexViewPos.xyz), dFdy(vertexViewPos.xyz));
     mat2   screenToUv   = mat2  (dFdx(gl_TexCoord[0].xy), dFdy(gl_TexCoord[0].xy)) / TEXTURE_OFFSET;
-    //mat2 uvToScreem = inverse(screenToUv);
-    //inverse is not available in GLSL 1.20, so calculate it manually:
-        float det = screenToUv[0][0] * screenToUv[1][1] - screenToUv[1][0] * screenToUv[0][1];
-        mat2 uvToScreen = mat2(vec2(screenToUv[1][1],-screenToUv[0][1]),vec2(-screenToUv[1][0],screenToUv[0][0])) / det;
+    mat2 uvToScreen = inverse2(screenToUv);
     mat2x3 uvToView = screenToView * uvToScreen;
-    mat3 tbn = mat3(uvToView[0], uvToView[1], normal);
 
 #if defined (PARALLAX_MAPPING)
-    vec3 eyeTangentSpace = normalizedVPos * tbn;
+    vec2 viewDirectionTextureOffset = normalizedVPos * uvToView;
```

Comment:
What's normalized-V-Pos? Is it normalized-Vertex-Position? And "normalized" in the sense of  Normalized Device Coordinates?

And can you help me understand viewDirectionTextureOffset? When I think about the view I think about a camera/eye space, and when I think about viewDirection I think about the single vector going out of the lens of the camera - neither, I feel, are the correct interpretation here, which hints to an ambiguity in the naming. Can we try and resolve this?

Reply:
That second interpretation, the single vector going out of the lens of the camera, is exactly what I meant by that, unless I'm not understanding what you mean by that phrase. Would viewDirectionUvProjection be better? I can't quite call it the view direction in UV coordinates, as UV coordinates don't span the whole view-space, but it is the perpendicular projection onto that.

Reply:
So it's the forward axis of the view space projected onto the UV space? If that's the case, and I know you don't like this kind of mouthfuls, I'd call it viewForwardAxisProjectedOntoUVspace. Sadly it's the only way to be unambiguous about it.

And I remind you my question: _What's normalized-V-Pos? Is it normalized-Vertex-Position? And "normalized" in the sense of Normalized Device Coordinates?_

Reply:
It's normalized as in same direction but unit length.